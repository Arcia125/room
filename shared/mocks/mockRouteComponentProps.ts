/**
 * @description Mock props for Children of Route components.
 */
import { RouteComponentProps, match } from 'react-router';
import * as history from 'history';

const mockLocation: history.Location = {
  hash: '',
  pathname: '/',
  search: '/',
  state: {},
};

const mockHistory: history.History = {
  push: jest.fn(),
  length: 1,
  block: jest.fn(),
  action: 'PUSH',
  createHref: jest.fn(),
  go: jest.fn(),
  goBack: jest.fn(),
  goForward: jest.fn(),
  listen: jest.fn(),
  location: mockLocation,
  replace: jest.fn(),
};

const mockMatch: match = {
  isExact: true,
  params: {},
  path: '/',
  url: '/',
};

/**
 * @description mocked version of props passed from react router Routes to their children
 * @example
 * render(<MyComponent {...mockRouteComponentProps} />)
 */
export const mockRouteComponentProps: RouteComponentProps = {
  history: mockHistory,
  location: mockLocation,
  match: mockMatch,
};
