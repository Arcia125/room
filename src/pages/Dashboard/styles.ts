import styled from 'styled-components';

const StyledDashboard = styled.div`
  display: grid;
  grid-template-rows: 100vh;
  grid-template-columns: 192px 1fr;
  & main {
    background-color: ${p => p.theme.colors.lighterPurple};
    padding-left: 39px;
    padding-top: 25px;
  }
`;

export { StyledDashboard };
