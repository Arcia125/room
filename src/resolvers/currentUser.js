import { getUser, setUser, getToken } from '../utils/user';
// let user = {
//   id: null,
//   username: null,
//   __typename: 'User'
// };

// let user = getUser();

const currentUser = {
  defaults: {
    currentUser: getUser()
  },
  resolvers: {
    Mutation: {
      saveUser: (_, { username, id, email, token }, { cache }) => {
        const user = { id, username, email, __typename: 'User' };
        console.log('saveUser resolver', user, token);
        // console.log('saving user ', user);
        // cache.writeQuery({
        //   query: GET_CURRENT_USER,
        //   data: { currentUser: user },
        // });

        setUser({ user, cache, token });

        return { user, token, __typename: 'AuthPayload' };
        // cache.writeData({
        //   currentUser: {
        //     id,
        //     username,
        //     __typename: 'User',
        //   },
        // });
      }
    },
    Query: {
      currentUser: (root, args, { cache }, info) => {
        // debugger;
        console.log('getting current user');
        // const queryResult = cache.readQuery({
        //   query: GET_CURRENT_USER,
        // });
        // console.log('readQuery', queryResult);
        const user = getUser();
        return user ? user : null;
        // return {
        //   user: getUser(),
        //   token: getToken(),
        // };
      }
    }
  }
};

export { currentUser };
