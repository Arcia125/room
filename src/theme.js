const theme = {
  spacing: {
    0: '0px',
    1: '8px',
    2: '16px',
    3: '32px'
  }
};

export { theme };
