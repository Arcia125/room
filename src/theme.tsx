import React, { ReactElement } from 'react';
import { ThemeProvider } from 'styled-components';

const theme = {
  spacing: {
    0: '0px',
    1: '8px',
    2: '16px',
    3: '32px',
  },
  colors: {
    gray: '#B3B2C0',
    grayDark: '#1f1f1f',
    lightBlue: '#AEAFE8',
    blue: '#AC2FC9',
    lightPurple: '#EFEEF8',
    purple: '#9056D4',
    pink: '#DD6589',
    bleh: '#8A4966',
  },
};

const Provider: React.FunctionComponent<{ children: ReactElement }> = ({
  children,
  ...restProps
}) => (
  <ThemeProvider theme={theme} {...restProps}>
    {children}
  </ThemeProvider>
);

export { theme, Provider };
