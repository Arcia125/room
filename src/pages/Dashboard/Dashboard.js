import React from 'react';
import PropTypes from 'prop-types';
import { useQuery, useMutation } from '@apollo/react-hooks';

import { GET_CURRENT_USER } from '../../graphql/getCurrentUser';
import { ADD_ROOM } from '../../graphql/addRoom';
import { Button } from '../../components/styles/Button';

const roomNames = ['test1', 'test2'];

const Dashboard = ({ history }) => {
  const getCurrentUserQuery = useQuery(GET_CURRENT_USER);

  const [addRoom, addRoomMutation] = useMutation(ADD_ROOM, {
    variables: { name: roomNames[Math.random() > 0.5 ? 0 : 1] }
  });

  const addedRoomId = addRoomMutation.data && addRoomMutation.data.addRoom.id;

  React.useEffect(() => {
    if (addedRoomId || addedRoomId === 0) {
      history.push(`/r/${addedRoomId}`);
    }
  });

  const handleCreateRoom = () => {
    addRoom();
  };

  if (getCurrentUserQuery.loading) return 'Loading...';
  if (getCurrentUserQuery.error) return getCurrentUserQuery.error.message;
  return (
    <div>
      <h1>
        {getCurrentUserQuery.data &&
          getCurrentUserQuery.data.currentUser &&
          getCurrentUserQuery.data.currentUser.username}
        's dashboard
      </h1>
      <p>Create a room to get started</p>
      <Button color="primary" padding="2" onClick={handleCreateRoom}>
        {getCurrentUserQuery.loading || addRoomMutation.loading
          ? 'loading...'
          : 'create a room'}
      </Button>
    </div>
  );
};

Dashboard.propTypes = {};

export default Dashboard;
