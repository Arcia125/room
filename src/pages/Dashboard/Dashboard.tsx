import React from 'react';
import PropTypes from 'prop-types';
import { useQuery, useMutation } from '@apollo/react-hooks';
import { RouteComponentProps } from 'react-router';

import { GET_CURRENT_USER } from '../../graphql/getCurrentUser';
import { ADD_ROOM } from '../../graphql/addRoom';
import { Button } from '../../components/styles/Button';
import { ClaimAccountForm } from '../../components/ClaimAccountForm';
import { CLAIM_ACCOUNT } from '../../graphql/claimAccount';
import Modal from '../../components/Modal';
import { SAVE_USER } from '../../graphql/saveUser';

const roomNames = ['test1', 'test2'];

const useModalState = (
  { initialIsOpen = false } = { initialIsOpen: false }
) => {
  const [isOpen, setIsOpen] = React.useState(initialIsOpen);

  const open = () => setIsOpen(true);

  const close = () => setIsOpen(false);

  return {
    isOpen,
    open,
    close
  };
};

const Dashboard: React.FunctionComponent<RouteComponentProps> = ({
  history
}) => {
  const getCurrentUserQuery = useQuery(GET_CURRENT_USER);

  const claimAccountModal = useModalState();

  const [saveUser, saveUserMutation] = useMutation(SAVE_USER);

  const [claimAccount, claimAccountMutation] = useMutation(CLAIM_ACCOUNT, {
    update: (client, mutResult) => {
      const variables = {
        ...mutResult.data.claimAccount.user,
        token: mutResult.data.claimAccount.token
      };

      console.log('saving user ', variables);

      saveUser({
        variables
      });

      window.location.reload();
    }
  });

  const [addRoom, addRoomMutation] = useMutation(ADD_ROOM, {
    variables: { name: roomNames[Math.random() > 0.5 ? 0 : 1] }
  });

  const addedRoomId = addRoomMutation.data && addRoomMutation.data.addRoom.id;

  React.useEffect(() => {
    if (addedRoomId || addedRoomId === 0) {
      history.push(`/r/${addedRoomId}`);
    }
  });

  if (getCurrentUserQuery.loading) return <div>'Loading...'</div>;
  if (getCurrentUserQuery.error)
    return <div>{getCurrentUserQuery.error.message}</div>;

  const currentUser =
    getCurrentUserQuery.data && getCurrentUserQuery.data.currentUser;

  const handleCreateRoom = () => {
    addRoom();
  };

  const handleClickMyAccount = () => {
    if (!currentUser.email) return claimAccountModal.open();
    history.push('/account');
  };

  return (
    <div>
      <h1>
        {getCurrentUserQuery.data &&
          getCurrentUserQuery.data.currentUser &&
          getCurrentUserQuery.data.currentUser.username}
        's dashboard
      </h1>
      <p>Create a room to get started</p>
      <Button color="primary" padding="2" onClick={handleCreateRoom}>
        {getCurrentUserQuery.loading || addRoomMutation.loading
          ? 'loading...'
          : 'create a room'}
      </Button>

      <Button color="primary" padding="2" onClick={handleClickMyAccount}>
        My Account
      </Button>

      <Modal
        isOpen={claimAccountModal.isOpen}
        onRequestClose={claimAccountModal.close}
      >
        <ClaimAccountForm onSubmit={variables => claimAccount({ variables })}>
          <Button type="submit" disabled={claimAccountMutation.loading}>
            Submit
          </Button>
        </ClaimAccountForm>
      </Modal>
    </div>
  );
};

Dashboard.propTypes = {};

export default Dashboard;
