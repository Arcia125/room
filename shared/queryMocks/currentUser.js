import { users } from '../mockData/users';
import { GET_CURRENT_USER } from '../../src/graphql/getCurrentUser';
import { currentUser } from '../../src/resolvers/currentUser';


const mockUser = users[0];

const currentUserMock = {
  request: {
    query: GET_CURRENT_USER,
  },
  result: {
    data: {
      currentUser: mockUser
    }
  }
};

const currentUserMockResolvers = {
  Mutation: currentUser.resolvers.Mutation,
  Query: {
    currentUser: () => mockUser
  }
};

const nullCurrentUserMock = {
  request: {
    query: GET_CURRENT_USER,
  },
  result: {
    data: {
      currentUser: null
    }
  }
};

const nullCurrentUserMockResolvers = {
  Mutation: currentUser.resolvers.Mutation,
  Query: {
    currentUser: () => null
  }
};



export { currentUserMock, currentUserMockResolvers, nullCurrentUserMock, nullCurrentUserMockResolvers };