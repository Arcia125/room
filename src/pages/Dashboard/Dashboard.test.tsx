import React from 'react';
import { MockedProvider } from '@apollo/react-testing';
import {
  render,
  cleanup,
  wait,
  fireEvent,
  waitForElement,
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import Dashboard from '.';
import { Provider } from '../../theme';
import { mockRouteComponentProps } from '../../../shared/mocks/mockRouteComponentProps';
import { currentUserMockResolvers } from '../../../shared/queryMocks/currentUser';
import { addRoomMock } from '../../../shared/queryMocks/room';
import { MemoryRouter } from 'react-router';

beforeEach(() => {
  jest.resetAllMocks();
});

afterEach(cleanup);

describe('Dashboard', () => {
  it('renders without crashing', async () => {
    const { container, asFragment } = render(
      <MemoryRouter>
        <MockedProvider addTypename resolvers={currentUserMockResolvers}>
          <Provider>
            <Dashboard {...mockRouteComponentProps} />
          </Provider>
        </MockedProvider>
      </MemoryRouter>
    );
    await wait(() => expect(container).toBeInTheDocument());

    expect(asFragment()).toMatchSnapshot();
  });

  it('allows the user to create a room', async () => {
    const { findByTestId } = render(
      <MemoryRouter>
        <MockedProvider
          addTypename
          mocks={[addRoomMock]}
          resolvers={currentUserMockResolvers}
        >
          <Provider>
            <Dashboard {...mockRouteComponentProps} />
          </Provider>
        </MockedProvider>
      </MemoryRouter>
    );

    const createRoomButton: HTMLElement = await waitForElement(() =>
      findByTestId('create-room-button')
    );

    fireEvent.click(createRoomButton);

    // after the createRoom mutation completes the component should redirect to the new room
    await wait(() =>
      expect(mockRouteComponentProps.history.push).toBeCalledTimes(1)
    );
  });
});
