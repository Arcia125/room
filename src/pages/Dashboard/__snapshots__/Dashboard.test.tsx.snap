// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard allows the user to create a room 1`] = `<DocumentFragment />`;

exports[`Dashboard renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa cdwebl"
  >
    <nav
      class="sc-bwzfXH kCtIxr"
    >
      <div
        class="user"
      >
        <img
          alt="avatar"
          src="https://source.unsplash.com/cqraK2a3Or8/50x50"
          style="border-radius: 50%;"
        />
        <p>
          Kevin
        </p>
      </div>
      <ul>
        <li>
          <button
            class="sc-htpNat klOdBn"
            data-testid="create-room-button"
          >
            <svg
              fill="none"
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 20V14H14V20H19V12H22L12 3L2 12H5V20H10Z"
                fill="#AEB3C8"
              />
            </svg>
            create room
          </button>
        </li>
        <li>
          <button
            class="sc-htpNat klOdBn"
          >
            <svg
              fill="none"
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19.14 12.936C19.176 12.636 19.2 12.324 19.2 12C19.2 11.676 19.176 11.364 19.128 11.064L21.156 9.48002C21.336 9.33602 21.384 9.07202 21.276 8.86802L19.356 5.54402C19.236 5.32802 18.984 5.25602 18.768 5.32802L16.38 6.28802C15.876 5.90402 15.348 5.59202 14.76 5.35202L14.4 2.80802C14.364 2.56802 14.16 2.40002 13.92 2.40002H10.08C9.83998 2.40002 9.64799 2.56802 9.61199 2.80802L9.25199 5.35202C8.66398 5.59202 8.12399 5.91602 7.63199 6.28802L5.24398 5.32802C5.02798 5.24402 4.77598 5.32802 4.65598 5.54402L2.73598 8.86802C2.61598 9.08402 2.66398 9.33602 2.85598 9.48002L4.88398 11.064C4.83598 11.364 4.79998 11.688 4.79998 12C4.79998 12.312 4.82398 12.636 4.87198 12.936L2.84398 14.52C2.66398 14.664 2.61598 14.928 2.72398 15.132L4.64398 18.456C4.76398 18.672 5.01598 18.744 5.23198 18.672L7.61998 17.712C8.12398 18.096 8.65198 18.408 9.23998 18.648L9.59999 21.192C9.64798 21.432 9.83998 21.6 10.08 21.6H13.92C14.16 21.6 14.364 21.432 14.388 21.192L14.748 18.648C15.336 18.408 15.876 18.084 16.368 17.712L18.756 18.672C18.972 18.756 19.224 18.672 19.344 18.456L21.264 15.132C21.384 14.916 21.336 14.664 21.144 14.52L19.14 12.936ZM12 15.6C10.02 15.6 8.39998 13.98 8.39998 12C8.39998 10.02 10.02 8.40002 12 8.40002C13.98 8.40002 15.6 10.02 15.6 12C15.6 13.98 13.98 15.6 12 15.6Z"
                fill="#AEB3C8"
              />
            </svg>
            account
          </button>
        </li>
        <li>
          <button
            class="sc-htpNat klOdBn"
          >
            <svg
              fill="none"
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18 1.01L8 1C6.9 1 6 1.9 6 3V6H8V5H18V19H8V18H6V21C6 22.1 6.9 23 8 23H18C19.1 23 20 22.1 20 21V3C20 1.9 19.1 1.01 18 1.01ZM10 15H12V8H5V10H8.59L3 15.59L4.41 17L10 11.41V15Z"
                fill="#AEB3C8"
              />
            </svg>
            logout
          </button>
        </li>
      </ul>
    </nav>
    <main>
      <a
        href="/"
      >
        <img
          alt="Room Logo"
          src="logoText.svg"
        />
      </a>
      <h1>
        Kevin's dashboard
      </h1>
      <p>
        Create a room to get started
      </p>
    </main>
  </div>
</DocumentFragment>
`;
